File: DSA/done/2 - addTwoNumbers.py
# https://leetcode.com/problems/add-two-numbers/

from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        dummy = res = ListNode()
        while l1 or l2 or carry:
            left = l1.val if l1 else 0
            right = l2.val if l2 else 0

            sum = left + right + carry
            carry = 1 if sum > 9 else 0
            sum %= 10

            dummy.next = ListNode(sum)
            dummy = dummy.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            
        return res.next


        

File: DSA/done/2 - addTwoNumbers.cpp
// https://leetcode.com/problems/add-two-numbers/

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry{0};
        ListNode* dummy = new ListNode();
        ListNode* res = dummy;
        while (l1 || l2 || carry) {
            int lvalue = l1 ? l1->val : 0;
            int rvalue = l2 ? l2->val : 0;

            int sum{ lvalue + rvalue + carry };
            carry = sum / 10;   
            sum %= 10;      

            dummy->next = new ListNode(sum);
            dummy = dummy->next;
            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        }
        return res->next;
    }
};